File: fui_system/src/platform/qt/qt_wrapper/cpp/qicon.cpp
#include <QIcon>
#include "qicon.h"

void *QIcon_new()
{
    return static_cast<void *>(new (std::nothrow) QIcon());
}

void QIcon_delete(void *self)
{
    delete static_cast<QIcon *>(self);
}

void QIcon_addPixmap(void *self, const void *pixmap, int mode, int state)
{
    QIcon *qIcon = static_cast<QIcon *>(self);
    const QPixmap *qPixmap = static_cast<const QPixmap *>(pixmap);

    qIcon->addPixmap(*qPixmap, (QIcon::Mode)mode, (QIcon::State)state);
}

File: fui_system/src/platform/qt/qt_wrapper/qicon.rs
use crate::platform::qt::qt_wrapper::QPixmap;
use crate::FUISystemError;

pub struct QIcon {
    pub this: *mut ::std::os::raw::c_void,
}

impl QIcon {
    pub fn new() -> Result<Self, FUISystemError> {
        unsafe {
            let this = crate::platform::qt::qt_wrapper::QIcon_new();
            if this.is_null() {
                return Err(FUISystemError::OutOfMemory);
            }

            Ok(Self { this })
        }
    }

    pub fn add_pixmap(&mut self, pixmap: &QPixmap) {
        unsafe {
            crate::platform::qt::qt_wrapper::QIcon_addPixmap(self.this, pixmap.this, 0, 1);
        }
    }
}

impl Drop for QIcon {
    fn drop(&mut self) {
        unsafe {
            crate::platform::qt::qt_wrapper::QIcon_delete(self.this);
        }
    }
}
